homeassistant/helpers/entity.py

```diff
@@ -259,7 +261,130 @@ class CalculatedState:
     shadowed_attributes: Mapping[str, Any]
 
 
-class Entity(ABC):
+class CachedProperties(type):
+    """Metaclass which invalidates cached entity propertis on write to _attr_."""
+
+    def __new__(
+        mcs,  # noqa: N804  ruff bug, ruff does not understand this is a metaclass
+        name: str,
+        bases: tuple[type, ...],
+        namespace: dict[Any, Any],
+        cached_properties: set[str] | None = None,
+        **kwargs: Any,
+    ) -> Any:
+        """Pop cached_properties and store it in the namespace."""
+        namespace["_CachedProperties__cached_properties"] = cached_properties or set()
+        return super().__new__(mcs, name, bases, namespace)
+
+    def __init__(
+        cls,
+        name: str,
+        bases: tuple[type, ...],
+        namespace: dict[Any, Any],
+        **kwargs: Any,
+    ) -> None:
+        """Wrap _attr_ for cached properties in property objects."""
```

Comment:
Add explanations to the docstring and maybe online comments about the logic here. What is the main goal, what approach is taken and details needed to make it work.